EngineLayer/FdrAnalysis/PEPValueAnalysisGeneric.cs

```diff
@@ -203,10 +211,45 @@ private static int GetChargeStateMode(List<PeptideSpectralMatch> psms)
                     }
                 }
 
+                List<double> allStDevs = new List<double>();
+
                 foreach (int key in hydrobophobicites.Keys)
                 {
                     //TODO consider using inner-quartile range instead of standard deviation
-                    averagesCommaStandardDeviations.Add(key, new Tuple<double, double>(hydrobophobicites[key].Average(), hydrobophobicites[key].StandardDeviation()));
+                    double averageHydrophobicity = hydrobophobicites[key].Average();
+                    averagesCommaStandardDeviations.Add(key, new Tuple<double, double>(averageHydrophobicity, hydrobophobicites[key].StandardDeviation()));
+                    foreach (double hydrophobicity in hydrobophobicites[key])
+                    {
+                        double difference = Math.Abs(hydrophobicity - averageHydrophobicity);
+                        if (!difference.Equals(double.NaN) && difference > 0)
+                        {
+                            allStDevs.Add(Math.Pow(difference, 2));
+                        }
+                    }
+                }
+
+                //some standard deviations are too small or too large because of random reasons, so we replace those small numbers of oddballs with reasonable numbers.
+                double globalStDev = 1;
+                if (allStDevs.Count() > 1)
+                {
+                    globalStDev = Math.Sqrt(allStDevs.Average());
+                }
+
+                Dictionary<int, Tuple<double, double>> stDevsToChange = new Dictionary<int, Tuple<double, double>>();
+                foreach (KeyValuePair<int, Tuple<double, double>> item in averagesCommaStandardDeviations)
+                {
+                    //add stability. not allowing stdevs that are too small or too large at one position relative to the global stdev
+                    //here we are finding which stdevs are out of whack.
+                    if (Double.IsNaN(item.Value.Item2) || item.Value.Item2 < 0.5 || (item.Value.Item2 / globalStDev) > 3)
```

Comment:
I don't think globalStDev is being calculated properly. line 235 is the global difference in hydrophobicity and bin hydrophobicity; I don't think it's been z-transformed